### PR TITLE
#658 Allow Create functionality [WIP]

### PR DESCRIPTION
**Pull request by vickyonit**
### Fix for following issues:

https://github.com/JedWatson/react-select/issues/658
https://github.com/JedWatson/react-select/issues/586
### Tech Changes:
- Modifies `Option` to render the "Add " label field if option is create
- Modifies `renderMenu` to show the new `Option` if `allowCreate` in the list
- Modifies Select Test Cases - Removed redundant tests
### Assumptions:
-  `addLabelText` contains `{label}` in it.
-  `newOptionCreator` always contains the following structure 
  `{
           value: value,
           label: value,
           create: true
  }`

### CODE REVIEW

**Comment by hsrobflavorus** [comment]
Thanks, I'm testing this out... the newOptionCreator missing is a big problem for us.

File: examples/src/components/AllowCreate.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	},
+
+	render () {
+		return (
+			<div className="section">
+				<h3 className="section-heading">{this.props.label}</h3>
+				<Select
+					allowCreate={this.props.allowCreate}
+					value={this.state.value}
+					multi={true}
```
Comment (omgaz) [inline examples/src/components/AllowCreate.js]: Generally preferred that boolean attributes are left like so `multi` e.g. https://github.com/JedWatson/react-select/blob/master/examples/src/components/NumericSelect.js#L66

File: src/Option.js [hunk 2]
```diff
@@ -11,6 +11,7 @@ const Option = React.createClass({
 		onFocus: React.PropTypes.func,                 // method to handle mouseEnter on option element
 		onUnfocus: React.PropTypes.func,               // method to handle mouseLeave on option element
 		option: React.PropTypes.object.isRequired,     // object that is base for that option
+		addLabelText: React.PropTypes.string,		   // text to display with value while creating new option
```
Comment (omgaz) [inline src/Option.js]: Minor fix, can you align the comment with the lines above.

File: examples/src/components/AllowCreate.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+					placeholder="Select your favourite(s)"
+					options={this.state.options}
+					onChange={this.handleSelectChange} />
+				{this.renderHint()}
+			</div>
+		);
+	}
+});
+
+module.exports = AllowCreate;
```
Comment (omgaz) [inline examples/src/components/AllowCreate.js]: New lines at the end of the file please :)

File: src/Select.js [hunk 4]
```diff
@@ -606,12 +616,15 @@ const Select = React.createClass({
...
+						addLabelText={this.props.addLabelText}
 						>
 						{renderLabel(option)}
 					</Option>
 				);
 			});
-		} else {
+		} else if (this.props.allowCreate){
+
+		} else{
```
Comment (omgaz) [inline src/Select.js]: One more space `} else {`

File: test/Select-test.js [hunk 5]
```diff
@@ -1047,7 +1047,7 @@ describe('Select', () => {
 	describe('with allowCreate=true', () => {
 
 		// TODO: allowCreate hasn't been implemented yet in 1.x
-		return;
+		// return;
```
Comment (omgaz) [inline test/Select-test.js]: I guess you can just remove the line completely, and the TODO comment as well :)

File: examples/src/components/AllowCreate.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+					allowCreate={this.props.allowCreate}
+					value={this.state.value}
+					multi={true}
+					placeholder="Select your favourite(s)"
+					options={this.state.options}
+					onChange={this.handleSelectChange} />
+				{this.renderHint()}
+			</div>
+		);
+	}
```
Comment (omgaz) [inline examples/src/components/AllowCreate.js]: ES6 extra comma `},`

**Comment by omgaz** [comment]
Great progress :)

**Comment by FoxxMD** [comment]
Looking forward to this!

**Comment by miro** [comment]
^ mandatory +1

**Comment by looshi** [comment]
+1

**Comment by feychenie** [comment]
:+1:

**Comment by RobAWilkinson** [comment]
:+1: this no allowCreate is a huge problem for us

**Comment by mitani** [comment]
This was a show stopper for use, but there is a workaround. For those of you that need allowCreate today - you can just use filterOptions and add an {label: `Add ${term}`, value: term, create:true} and then you just process it in your onChange function where you look for create === true

**Comment by donbonifacio** [comment]
:+1:

**Comment by dcwither** [comment]
:+1:

**Comment by throoze** [comment]
Looking forward to this being implemented in the new version. Thanks for your efforts guys!

**Comment by mehcode** [comment]
Definitely excited about this :100:

**Comment by omgaz** [comment]
LGTM :fish_cake:

**Comment by Craga89** [comment]
Is this on the roadmap to be merged soon? Looking forward to using it in production! :+1:

**Comment by wizhippo** [comment]
+1

**Comment by gfeld** [comment]
+1

**Comment by lefnire** [comment]
Thanks @mitani for the workaround! For anyone looking for more details on said, here's a [sample commit](https://github.com/lefnire/jobpig/commit/ce4bf390f47f41683f1d253c28dc748b14287d67). I like this solution as it grants more flexibility, and works for `multi={true|false}` where `allowCreate` only works for `multi={true}`. You can see how it looks at jobpigapp.com -> register -> post job (just play with the location & tags).

Incidentally this PR is getting pretty old. I tried to fork & rebase but was a bit out of my league. I have to ask: is the only reason this hasn't been merged cosmetic concerns (aligning comments, `} else {` issues, etc)? That'd be a shame :/ there are some people hurting for a merge...

**Comment by lefnire** [comment]
If it is rebase + style, I'd be happy to take a pass this week. If it's more, let us know

**Comment by cema-sp** [comment]
Thank you, @lefnire and @mitani !
For those who use ES6 without lodash:
~~~es6
  filterOptions = (options, filter, values) => {
    // Filter already selected values
    let filteredOptions = options.filter(option => {
      return !(values.includes(option));
    });

    // Filter by label
    if (filter !== undefined && filter != null && filter.length > 0) {
      filteredOptions = filteredOptions.filter(option => {
        return RegExp(filter, 'ig').test(option.label);
      });
    }

    // Append Addition option
    if (filteredOptions.length == 0) {
      filteredOptions.push({
        label:  <span><strong>Create</strong>: {filter}</span>,
        value:  filter,
        create: true,
      });
    }

    return filteredOptions;
  };
~~~

**Comment by JuanCrg90** [comment]
+1

**Comment by bIgBV** [comment]
Hey, I also want to add a +1 to this PR. 

If it is only stylistic changes stopping from this being merged, I can get started on  fixing them. We are using an older version of the library (0.9 to be specific) because it has the `allowCreate` option enabled. This is stopping us from upgrading to react v15.

**Comment by env** [comment]
+1

**Comment by sethmcleod** [comment]
Thank you @cema-sp for that ES6 solution! One thing I noticed is that you can change the last conditional statement to this: `filteredOptions.length === 0 && filter.length > 0` and it will prevent the _Create_ label from showing if the user has selected all options but hasn't actually typed anything.

Successfully using this in place of `allowCreate` for v1-beta in a react v15 build (with redux-form) and it's working great. Thanks to everyone who contributed!

**Comment by kopaygorodsky** [comment]
+1

**Comment by mehcode** [comment]
https://github.com/mehcode/react-select

^ I have no recollection of what I did but I _think_ I based it off of this PR and did _something_. It gives me warnings on npm about missing peer dependencies but it does work with React 15.x

**Comment by wbecker** [comment]
This looks like it has stalled since December - are there only those few little stylistic comments in the PR holding it back? I could quite happily do these if it meant that this would get merged ASAP @omgaz!

File: test/Select-test.js [hunk 7]
```diff
@@ -1047,7 +1047,7 @@ describe('Select', () => {
 	describe('with allowCreate=true', () => {
 
 		// TODO: allowCreate hasn't been implemented yet in 1.x
```
Comment (omgaz) [inline test/Select-test.js]: remove comment

File: src/Select.js [hunk 8]
```diff
@@ -606,12 +616,15 @@ const Select = React.createClass({
...
 						ref={optionRef}
+						addLabelText={this.props.addLabelText}
 						>
 						{renderLabel(option)}
 					</Option>
 				);
 			});
-		} else {
+		} else if (this.props.allowCreate){
+
```
Comment (omgaz) [inline src/Select.js]: What needs to go in here?

**Comment by omgaz** [comment]
THere's also a big missing chunk in an else if (this.props.allowCreate) {}, I'm not sure what vickyonit intended for this. It'd be awesome if someone would look into reviving this :)

**Comment by gwyneplaine** [comment]
Thanks @vickyonit, this has since been superseded by https://github.com/JedWatson/react-select/pull/1187 which abstracts option creation to a Creatable HoC.

**Pull request closed without merging**